[otlp exporter] Validate exporter endpoint has port

atmask opened the pull request:
**Description:** This PR updates the otlp exporter config validation to ensure that the "endpoint" specified for the exporter includes a port. The goal of this is to fail fast if the configuration is invalid instead of waiting for an error to arise. The PR adds a function to the ClientConfig defined in configgrpc that parses the port defined in the endpoint. The otlp exporter uses this port parsing to validate that 

**Link to tracking Issue:** [Issue 9505](https://github.com/open-telemetry/opentelemetry-collector/issues/9505)

**Testing:** 

configgrpc:
- Add unit tests to validate ports are correctly parsed in configrpc and that an error is returned if the port is omitted

otlpexporter:
- Add unit test for conifg validation function to ensure that the validation fails for missing endpoint or missing port and that a well-formed endpoint does not fail

TylerHelmuth commented on exporter/otlpexporter/config.go:
```diff
@@ -22,10 +22,31 @@ type Config struct {
 }
 
 func (c *Config) Validate() error {
+	if err := c.validateEndpointNotEmpty(); err != nil {
+		return err
+	}
+	if err := c.validateEndpointHasPort(); err != nil {
+		return err
+	}
```
The logic inside these functions is simple enough that we can keep everything inside `Validate`.

atmask replied:
Addressed in ec2b9d32

TylerHelmuth commented on config/configgrpc/configgrpc.go:
```diff
@@ -172,6 +172,16 @@ func (gcs *ClientConfig) SanitizedEndpoint() string {
 	}
 }
 
+func (gcs *ClientConfig) EndpointPort() (string, error) {
```
We're working towards GA for `configgrpc` right now and I am hesitant to add new public APIs. For now, let's move this function into the otlpexporter.

atmask replied:
Addressed in ec2b9d3

bogdandrutu commented on exporter/otlpexporter/config.go:
```diff
@@ -25,7 +27,26 @@ func (c *Config) Validate() error {
 	}
+
+	// Validate endpoint has a valid port
+	port, err := c.endpointPort()
```
We have a default port, why not using that if not provided?

TylerHelmuth commented on exporter/otlpexporter/config.go:
```diff
@@ -25,7 +27,23 @@ func (c *Config) Validate() error {
 }
 
+func (c *Config) endpointPort() (string, error) {
+	sanitizedEndpoint := c.SanitizedEndpoint()
+	_, port, err := net.SplitHostPort(sanitizedEndpoint)
```
```suggestion
	_, port, err := net.SplitHostPort(c.SanitizedEndpoint())
```

TylerHelmuth commented on exporter/otlpexporter/config.go:
```diff
@@ -25,7 +27,23 @@ func (c *Config) Validate() error {
 	return nil
 }
 
+func (c *Config) endpointPort() (string, error) {
```
@atmask ya that is normally the process - go update contrib and we'll get that merged to unblock this PR.

atmask replied:
PR for otel-collector-contrib linked below

TylerHelmuth commented on exporter/otlpexporter/config_test.go:
```diff
@@ -114,3 +114,26 @@ func TestUnmarshalInvalidConfig(t *testing.T) {
 
 }
+
+func TestConfigValidation(t *testing.T) {
```
Instead of a new function we normally test the config via `TestUnmarshalConfig`/`TestUnmarshalInvalidConfig`.  You can add new test scenarios to `testdata/` and then add scenarios to the appropriate test function.

atmask replied:
I see, that make sense. Moved the new test scenario into the testdata

atmask commented:
Related PR from otel-collector-contrib to update loadbalancingexporter for the change in otlp config validation: https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/31425/files

iblancasa commented on exporter/otlpexporter/testdata/invalid_configs.yaml:
```diff
@@ -53,3 +53,19 @@ invalid_retry:
     max_interval: 60s
     max_elapsed_time: 10m
+missing_port:
+  endpoint: example.com
```
Would it make sense to test an endpoint using a wrong port? Like `example.com:stuff` instead of `example.com:8080`

atmask replied:
Addressed in latest push. Resolving

atmask commented:
Just updated to fix the lint issue for new commits going into main and rebased on main. @TylerHelmuth could you please review this or poke a codeowner to push this PR through so that it can get closed off?

dmitryax commented:
I think using the default port is better, but we can add the validation for now since it's not working without specifying the port explicitly anyway. 

I think we don't have clear guidance on how components should mutate config structs passed to them when it's required, like in this case. It'd be great to have an issue for that so we can track it separately.
